### PR TITLE
feature: added TEST_NGINX_RAND_PORT_XXX variable to provide random ports during environment variable expansion.

### PR DESCRIPTION
**Pull request by xiaocang**

### CODE REVIEW

**Review by doujiang24 — APPROVED** [review]

**Review by agentzh — CHANGES_REQUESTED** [review]

File: lib/Test/Nginx/Util.pm [hunk 1]
```diff
@@ -78,6 +78,8 @@ our $ServerAddr = '127.0.0.1';
 
 our $ServerName = 'localhost';
 
+our %UsedPort;
```
[the remark this answers is not in the document]
Comment (xiaocang) [reply lib/Test/Nginx/Util.pm]: Okay, I've modified it that way.

File: lib/Test/Nginx/Util.pm [hunk 2]
```diff
@@ -174,6 +176,39 @@ sub gen_rand_str {
...
+
+        if (defined $sock) {
+            $sock->close();
+            $rand_port = $port;
+            $UsedPort{$port} = 1;
+            last;
+        }
+
+        if ($Verbose) {
+            warn "Try again, port $port is already in use: $@\n";
```
[the remark this answers is not in the document]
Comment (xiaocang) [reply lib/Test/Nginx/Util.pm:204]: Yeah, thanks for reminding me. I already added it.

File: lib/Test/Nginx/Socket.pm [hunk 3]
```diff
@@ -4448,6 +4448,14 @@ C<prove -j8 -r t> runs the test suite under F<t/> in 8 parallel jobs, utilizing
 Note that only test suite I<without> external shared and writable service dependencies (like Memcached,
 Redis or MySQL) can run in parallel in this way, obviously.
 
+=head2 TEST_NGINX_RANDOM_PORT[0-9]+
+
+Variables like C<TEST_NGINX_RANDOM_PORT[0-9]+> are expanded to random, unused
```
[the remark this answers is not in the document]
Comment (xiaocang) [reply lib/Test/Nginx/Socket.pm]: Ok, I've modified the document like this.

File: lib/Test/Nginx/Socket.pm [hunk 4]
```diff
@@ -4448,6 +4448,14 @@ C<prove -j8 -r t> runs the test suite under F<t/> in 8 parallel jobs, utilizing
 Note that only test suite I<without> external shared and writable service dependencies (like Memcached,
 Redis or MySQL) can run in parallel in this way, obviously.
 
+=head2 TEST_NGINX_RANDOM_PORT[0-9]+
```
[the remark this answers is not in the document]
Comment (xiaocang) [reply lib/Test/Nginx/Socket.pm]: I've added "Special Variables" to a separate section under "environment variables", you can help to see if it fits your description.

[the remark this answers is not in the document]
Comment (xiaocang) [reply lib/Test/Nginx/Socket.pm]: I've added "Special Variables" to a separate section under "environment variables", you can help to see if it fits your description.

File: lib/Test/Nginx/Socket.pm [hunk 5]
```diff
@@ -4462,6 +4462,18 @@ One can also add an optional user group separated by spaces, as in
...
 
+=head2 Special Variables
+
+=head3 TEST_NGINX_RAND_PORT_XXX
+
+Variables like C<TEST_NGINX_RAND_PORT_XXX> are expanded to random and unused
+unprivileged ports numbers to build the configuration of the server,
+where C<XXX> is an integer from 1985 to 65535.
+
+For instance, C<TEST_NGINX_RAND_PORT_1> will be expanded to an integer port,
```
Comment (agentzh) [inline lib/Test/Nginx/Socket.pm:4473]: “integer port" => "integer port number" (still thinking "integer" is redundant. is there any float-point number port numbers?)

File: lib/Test/Nginx/Util.pm [hunk 6]
```diff
@@ -174,6 +174,39 @@ sub gen_rand_str {
     return $s;
 }
 
+sub gen_rand_port (;$$) {
```
Comment (agentzh) [inline lib/Test/Nginx/Util.pm:177]: Here you assume the `tries` argument can be optional. But you never initialize it when it is not defined?

**Review by agentzh — CHANGES_REQUESTED** [review]

File: lib/Test/Nginx/Socket.pm (hunk 5 above)
Comment (xiaocang) [reply lib/Test/Nginx/Socket.pm]: ok, I've changed "integer port" to "port number".

File: lib/Test/Nginx/Util.pm (hunk 6 above)
Comment (xiaocang) [reply lib/Test/Nginx/Util.pm:177]: I missed it. It's been added.

**Pull request merged**